### DISCUSSION TITLE
Here are Zsh Completions for shell_gpt

### DISCUSSION THREAD

**Discussion by eiger8** [TheR1D/shell_gpt#468] [Show and tell]
Hello, sgpt community!

I'm excited to share developed Zsh completions for the sgpt, aiming to enhance our command-line experience. You can find mentioned modules in attached files.

**Key Features**

- flags and options `TAB` suggestion
- options `--model, --chat, --show-chat, --role, -- show-role` have interactive arguments suggestion (press TAB after option)
- support for 2 comletion engines:
     - plain zsh comletion
     - [argc-completions](https://github.com/sigoden/argc-completions)

**Installation** 

1. Plain zsh comletion:
    Copy the `_sgpt` into the default zsh or zsh's framework completion folders:
        - default one is `~/.zsh/completion`
        - Oh-My-Zsh framework's one `~/.oh-my-zsh/cache/completions`
2. argc-completions engine:
    Copy the `sgpt.sh` to the engine comletion folder `../argc-completions/completions`
        - if yo followed [installation instructions](https://github.com/sigoden/argc-completions?tab=readme-ov-file#getting-started) it should be here `$ARGC_COMPLETIONS_ROOT/completions`

**Usage**

Once installed, the completions will automatically activate as you type sgpt command follower by `--` and pressing `TAB` in your terminal aap . You'll see suggestions for command options and arguments, making it easier to construct your commands. 
NOTE: for the `--model` suggestion to work correctly you should have the env var `OPENAI_API_KEY` configured in `~/.config/shell_gpt/.sgptrc`

[sgpt-shell-comletions-v0.2.zip](https://github.com/TheR1D/shell_gpt/files/14209996/sgpt-shell-comletions-v0.2.zip)

**UPD with fix for '_sgpt' file so the role function will work corectly:**
[sgpt-shell-comletions-v0.3.zip](https://github.com/TheR1D/shell_gpt/files/14318310/sgpt-shell-comletions-v0.3.zip)

![iTerm2 2024-02-08 at 15 51 27](https://github.com/TheR1D/shell_gpt/assets/21951841/12f13c15-b2b2-4a1a-a8eb-7764aa5f0d26)

**Comment by eiger8** [comment]
I've uploaded archive with fix for the '_sgpt' completion script that should fix typo in for get-role list function.
[sgpt-shell-comletions-v0.3.zip](https://github.com/TheR1D/shell_gpt/files/14318310/sgpt-shell-comletions-v0.3.zip)

**Comment by ristomatti** [comment]
Thanks @eiger8! This inspired me to roll the dice to fix this nagging missing feature for me as well.

```sh
$ sgpt --code "Convert this zsh completion file to bash completion" < _sgpt | tee sgpt.bash
_sgpt()
{
    local cur prev opts
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    opts="--model --temperature --top-p --editor --no-editor --cache --no-cache --version --help --shell --interaction --no-interaction --describe-shell --code --functions --no-functions --chat --repl --show-chat --list-chats --role --create-role --show-role --list-roles --install-integration"

    case "${prev}" in
        --model)
            COMPREPLY=( $(compgen -W "$(curl -s https://api.openai.com/v1/models -H "Content-Type: application/json" -H "Authorization: Bearer $(grep '^OPENAI_API_KEY=' ~/.config/shell_gpt/.sgptrc | cut -d '=' -f2)" | jq -r '.data[].id')" -- "${cur}") )
            return 0
            ;;
        --chat|--show-chat)
            COMPREPLY=( $(compgen -W "$(sgpt --list-chats | sed -E 's|.*/chat_cache/(.*)|\1|')" -- "${cur}") )
            return 0
            ;;
        --role|--show-role)
            COMPREPLY=( $(compgen -W "$(sgpt --list-roles | sed -E 's|.*/roles/(.*)\.json|\1|')" -- "${cur}") )
            return 0
            ;;
        *)
            ;;
    esac

    if [[ ${cur} == -* ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
    fi
}

complete -F _sgpt sgpt
```

```sh


$ cp sgpt.bash ~/.local/share/bash-completion/completions/
```

**Result**

There's some things broken but this already saves a lot of typing!


https://github.com/TheR1D/shell_gpt/assets/9029939/6a5e5616-a9b7-4a62-9ab0-3842433d4c4f

**Reply by ristomatti** [reply]
Broken things e.g.

```sh
# Garbage output, although "temp" is an actual chat
$ sgpt --show-chat  # or --chat 
been            does            have            It              should          to
but             file?           How             not             support         temp

# Truncated option, easy to fx of course
$ sgpt --top-p
```

**Reply by eiger8** [reply]
@ristomatti did not get what missing feature you meant 🤔
In my case I have fzf enabled in my shell (oh-my-zsh + fzf + fzf-tab). This provides the same functionality as shown on your video and adds far more flexibility with fuzzy search for history and any suggestions.

**Reply by ristomatti** [reply]
@eiger8 I was missing bash completions. I've used it so long, I'm reluctant to move away. I've got fzf and fzf tab completion working but for this type of short lists I find the popup jarring. It's behind `ctrl-space`.

Rather it was a fun experiment to see what comes out. GPT seems to be quite good at crafting shell scripts. My latest revolution was to create a text file with a mockup of `some_command --help`, then pipe it to `sgpt` and ask to implement it and pipe the output to another file. I've created lots of nice to have utilities this way I otherwise wouldn't have time to do. Shell script syntax seems to never stick, I always need to google even the basic if statements.

**Reply by ristomatti** [reply]
@eiger8 After my reply I actually ended up setting up omz and converting most of my >700 line `.bashrc` over. I have to admit the `--help` style tab completion for params is pretty great at least. Great work with the `sgpt` completions, thanks!

**Reply by eiger8** [reply]
Good to hear that someone was inspired to move to zsh by me 😀

**Comment by eiger8** [comment]
Here is an updated reflecting changes in shell_gpt v1.4 (added --md, --no-md option)
[sgpt-shell-comletions-v0.4.zip](https://github.com/TheR1D/shell_gpt/files/14538540/sgpt-shell-comletions-v0.4.zip)

**Reply by ristomatti** [reply]
Say what, v1.4?

```sh
$ sgpt --version
ShellGPT 1.0.1
```

I've only installed it ~2 weeks ago. It appears `pipx` doesn't always have the newest release (or I'm doing it wrong).

**Reply by siikdUde** [reply]
<img width="134" alt="Screenshot 2" src="https://github.com/TheR1D/shell_gpt/assets/10148714/f8613f70-0c65-40c6-bbe7-a31e2ab85df9">


1.4.0 was the release... 2 weeks ago

**Reply by ristomatti** [reply]
It turned out I had to create a Python virtual environment with a newer Python version, then install it with `pipx reinstall --python <path_to_python_3.12> sgpt`. You learn something new every day.

```sh
$ sgpt --version
ShellGPT 1.4.0
``

**Reply by eiger8** [reply]
Run `pipx upgrade shell-gpt` to update sgpt in any installed venv
Or run `pipx upgrade-all` to update all the packages managed by `pipx`

**Note: pipx is managing all the venvs automatically. In most cases you do not need to create venvs manually.**

Everithing described above is valid in case pipx is installed correctly.

**Reply by ristomatti** [reply]
@eiger8 Thanks for the reply. That was my understanding as well. But since I had originally installed `shell_gpt` while I had the OS default Python 3.8 as the default, it had created a venv based on that. `pipx upgrade shell_gpt` just told it's the latest version. Anyhow, this was a useful finding since 3.11 apparently brought in some [non-neglible performance improvements](https://www.phoronix.com/review/python-311-performance). I had all Python apps using a 3.8 venv.
